[Docs] Add examples for new object-based c10d APIs

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#43932 [Docs] Add examples for new object-based c10d APIs**
* #48909 Add object-based collective APIs to public docs

Adds some basic examples to the documentation for each of the newly added
object-based collectives.

Differential Revision: [D23441838](https://our.internmc.facebook.com/intern/diff/D23441838/)

rohan-varma committed:
[Docs] Add examples for new object-based c10d APIs

Adds some basic examples to the documentation for each of the newly added
object-based collectibves.

Differential Revision: [D23441838](https://our.internmc.facebook.com/intern/diff/D23441838/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[Docs] Add examples for new object-based c10d APIs"


Adds some basic examples to the documentation for each of the newly added
object-based collectibves.

Differential Revision: [D23441838](https://our.internmc.facebook.com/intern/diff/D23441838/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[Docs] Add examples for new object-based c10d APIs"


Adds some basic examples to the documentation for each of the newly added
object-based collectives.

Differential Revision: [D23441838](https://our.internmc.facebook.com/intern/diff/D23441838/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[Docs] Add examples for new object-based c10d APIs"


Adds some basic examples to the documentation for each of the newly added
object-based collectives.

Differential Revision: [D23441838](https://our.internmc.facebook.com/intern/diff/D23441838/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[Docs] Add examples for new object-based c10d APIs"


Adds some basic examples to the documentation for each of the newly added
object-based collectives.

Differential Revision: [D23441838](https://our.internmc.facebook.com/intern/diff/D23441838/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[Docs] Add examples for new object-based c10d APIs"


Adds some basic examples to the documentation for each of the newly added
object-based collectives.

Differential Revision: [D23441838](https://our.internmc.facebook.com/intern/diff/D23441838/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[Docs] Add examples for new object-based c10d APIs"


Adds some basic examples to the documentation for each of the newly added
object-based collectives.

Differential Revision: [D23441838](https://our.internmc.facebook.com/intern/diff/D23441838/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[Docs] Add examples for new object-based c10d APIs"


Adds some basic examples to the documentation for each of the newly added
object-based collectives.

Differential Revision: [D23441838](https://our.internmc.facebook.com/intern/diff/D23441838/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[Docs] Add examples for new object-based c10d APIs"


Adds some basic examples to the documentation for each of the newly added
object-based collectives.

Differential Revision: [D23441838](https://our.internmc.facebook.com/intern/diff/D23441838/)

[ghstack-poisoned]

mrshenli commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1393,6 +1393,15 @@ def all_gather_object(object_list, obj, group=group.WORLD):
+        >>> output = [None for _ in gather_objects]
+        >>> dist.all_gather_object(output, gather_objects[dist.get_rank()])
+        >>> output
+        ['foo', 12, {1: 2}]
```
shall we verify if this line is rendered correctly?

mrshenli commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1467,6 +1476,20 @@ def gather_object(obj, object_gather_list=None, dst=0, group=group.WORLD):
+        >>> gather_objects = ["foo", 12, {1: 2}] # any picklable object
+        >>> output = [None for _ in gather_objects]
+        >>> dist.gather_object(
+                gather_objects[dist.get_rank()],
+                output if dist.get_rank() == 0 else None,
+                dst=0
+            )
```
shall we also verify if these lines are rendered correctly?

mrshenli commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1556,6 +1579,17 @@ def broadcast_object_list(object_list, src, group=group.WORLD):
+        >>> else:
+        >>>     objects = [None, None, None]
+        >>> dist.broadcast_object_list(objects, src=0)
+        >>> broadcast_objects
```
this var is not defined?

rohan-varma replied:
Good catch, it should be just `objects`.

mrshenli commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1393,6 +1393,15 @@ def all_gather_object(object_list, obj, group=group.WORLD):
+    Example::
+        >>> # Note: Process group initialization omitted on each rank.
+        >>> import torch.distributed as dist
+        >>> gather_objects = ["foo", 12, {1: 2}] # any picklable object
```
this assumes `world_size==3`?

rohan-varma replied:
Right, I guess we could just add this as a comment to keep the example clear and concise.

mrshenli commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1634,6 +1668,19 @@ def scatter_object_list(
+        >>> if dist.get_rank() == 0:
+        >>>     objects = ["foo", 12, {1: 2}] # any picklable object
+        >>> else:
+        >>>     objects = [None, None, None]
```
does this have to be 3 Nones?

rohan-varma replied:
It's mostly to point out that whatever is contained in `objects` on non-src ranks, it doesn't matter, we will only look at the `objects` from the src rank. The 2nd arg can be anything for non-src ranks, it is not considered.

mrshenli reviewed (approved):
LGTM! Thanks for adding this! Left some minor comments mainly about rendering and clarity.

rohan-varma committed:
Update on "[Docs] Add examples for new object-based c10d APIs"


Adds some basic examples to the documentation for each of the newly added
object-based collectives.

Differential Revision: [D23441838](https://our.internmc.facebook.com/intern/diff/D23441838/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[Docs] Add examples for new object-based c10d APIs"


Adds some basic examples to the documentation for each of the newly added
object-based collectives.

Differential Revision: [D23441838](https://our.internmc.facebook.com/intern/diff/D23441838/)

[ghstack-poisoned]

rohan-varma commented:
@mrshenli Looks like the rendering is as expected after building the docs locally, see for example: 

![all_gather_ss](https://user-images.githubusercontent.com/8039770/101396258-e6da4b80-387f-11eb-9887-59f1971f202c.png)
